src/multiplexed-session.ts

```diff
@@ -0,0 +1,320 @@
+/*!
+ * Copyright 2024 Google LLC. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {EventEmitter} from 'events';
+import PQueue from 'p-queue';
+import {Database} from './database';
+import {Session} from './session';
+import {Transaction} from './transaction';
+import {
+  isDatabaseNotFoundError,
+  isInstanceNotFoundError,
+  isCreateSessionPermissionError,
+  isDefaultCredentialsNotSetError,
+  isProjectIdNotSetInEnvironmentError,
+} from './session-pool';
+
+interface MuxSession {
+  multiplexedSession: Session | null;
+}
+
+/**
+ * @callback GetSessionCallback
+ * @param {?Error} error Request error, if any.
+ * @param {Session} session The read-write session.
+ * @param {Transaction} transaction The transaction object.
+ */
+export interface GetSessionCallback {
+  (
+    err: Error | null,
+    session?: Session | null,
+    transaction?: Transaction | null
+  ): void;
+}
+
+/**
+ * Interface for implementing multiplexed session logic, it should extend the
+ * {@link https://nodejs.org/api/events.html|EventEmitter} class
+ *
+ * @interface MultiplexedSessionInterface
+ * @extends external:{@link https://nodejs.org/api/events.html|EventEmitter}
+ *
+ * @constructs MultiplexedSessionInterface
+ * @param {Database} database The database to create a multiplexed session for.
+ */
+export interface MultiplexedSessionInterface {
+  /**
+   * When called creates a multiplexed session.
+   *
+   * @name MultiplexedSessionInterface#createSession
+   */
+  createSession(): void;
+
+  /**
+   * When called returns a multiplexed session.
+   *
+   * @name MultiplexedSessionInterface#getSession
+   * @param {GetSessionCallback} callback The callback function.
+   */
+  getSession(callback: GetSessionCallback): void;
+}
+
+/**
+ * Multiplexed session configuration options.
+ *
+ * @typedef {object} MultiplexedSessionOptions
+ * @property {number} [refreshRate=7] How often to check for expiration of multiplexed session, in
+ *     days. Must not be greater than 7 days.
+ */
+
+export interface MultiplexedSessionOptions {
```

Comment:
Yeah, in case of mux session we not exposing any options to our customers. I have removed the MultiplexedSessionOptions instead now I am using a refreshRate property inside the class and assign the default value to it in the constructor